File: ex-14/ex-14_java/Principal.java
import java.util.List;
import java.util.Random;
import java.util.Scanner;

public class Principal {
    private JanelaRender janela;
    private Simulacao simulacao;
    private Renderizador renderizador;

    private final int larguraJanela = 1200;
    private final int alturaJanela = 800;

    private Scanner scanner;
    private Random random;

    public Principal() {
        janela = new JanelaRender();
        simulacao = null;
        renderizador = null;
        scanner = new Scanner(System.in);
        random = new Random();
    }

    private double lerDouble() {
        return Double.parseDouble(scanner.next());
    }

    private int lerInt() {
        return Integer.parseInt(scanner.next());
    }

    private void inicializar() {
        janela.create(larguraJanela, alturaJanela, "Simulacao Gravitacional");
        janela.setFramerateLimit(60);
        obterDadosUsuario();
    }

    private void obterDadosUsuario() {
        System.out.println("=== SIMULACAO DE ATRACAO GRAVITACIONAL ===");
        System.out.println();

        double larguraEspaco, alturaEspaco;
        System.out.print("Digite a largura do espaco (em metros, ex: 100e9 para 100 milhoes de km): ");
        larguraEspaco = lerDouble();
        System.out.print("Digite a altura do espaco (em metros): ");
        alturaEspaco = lerDouble();

        double dt;
        System.out.print("Digite o intervalo de tempo (em segundos, ex: 3600 para 1 hora): ");
        dt = lerDouble();

        simulacao = new Simulacao(larguraEspaco, alturaEspaco, dt);
        renderizador = new Renderizador(janela, larguraEspaco, alturaEspaco);

        System.out.println();
        System.out.print("Deseja posicionar os corpos aleatoriamente? (1-Sim, 0-Nao): ");
        int aleatorio;
        aleatorio = lerInt();

        for (int i = 0; i < 2; i++) {
            System.out.println();
            System.out.println("=== CORPO " + (i + 1) + " ===");

            double posX, posY;
            if (aleatorio == 1) {
                posX = random.nextDouble() * larguraEspaco;
                posY = random.nextDouble() * alturaEspaco;
                System.out.println("Posicao aleatoria: X=" + posX + " m, Y=" + posY + " m");
            } else {
                System.out.print("Digite a posicao X (em metros): ");
                posX = lerDouble();
                System.out.print("Digite a posicao Y (em metros): ");
                posY = lerDouble();
            }

            double velX, velY;
            System.out.print("Digite a velocidade X (em m/s, ex: 30000 para 30 km/s): ");
            velX = lerDouble();
            System.out.print("Digite a velocidade Y (em m/s): ");
            velY = lerDouble();

            double massa;
            System.out.print("Digite a massa (em kg, ex: 6e27): ");
            massa = lerDouble();

            double raio = 10.0 + i * 5.0;
            Cor cor = (i == 0) ? Cor.Yellow() : Cor.Cyan();

            Vetor2D posicao = new Vetor2D(posX, posY);
            Vetor2D velocidade = new Vetor2D(velX, velY);

            CorpoCeleste corpo = new CorpoCeleste(posicao, velocidade, massa, raio, cor);
            simulacao.adicionarCorpo(corpo);
        }

        System.out.println();
        System.out.println("Simulacao iniciada! Pressione ESC para sair.");
        System.out.println("Pressione ESPACO para pausar/continuar.");
        System.out.println("Pressione R para reiniciar.");
    }

    private void processarEventos() {
        Evento evento;
        while ((evento = janela.pollEvent()) != null) {
            if (evento.isClosed()) {
                janela.close();
            }

            Integer keyPressed = evento.getKeyPressed();
            if (keyPressed != null) {
                if (keyPressed == Evento.Escape) {
                    janela.close();
                }
                if (keyPressed == Evento.R) {
                    finalizar();
                    obterDadosUsuario();
                }
            }
        }
    }

    private void atualizar() {
        if (simulacao != null) {
            simulacao.atualizar();

            for (int i = 0; i < simulacao.getCorpos().size(); i++) {
                CorpoCeleste corpo = simulacao.getCorpos().get(i);
                int[] tamanhoJanela = janela.getSize();

                double escalaX = tamanhoJanela[0] / simulacao.getLarguraEspaco();
                double escalaY = tamanhoJanela[1] / simulacao.getAlturaEspaco();

                float x = (float) (corpo.getPosicao().getX() * escalaX);
                float y = (float) (corpo.getPosicao().getY() * escalaY);

                Vetor2f posicaoTela = new Vetor2f(x, y);

                corpo.adicionarPontoTrajetoria(posicaoTela);
            }
        }
    }

    private void renderizar() {
        if (renderizador != null && simulacao != null) {
            renderizador.renderizar(simulacao);
        }
    }

    private void finalizar() {
        if (simulacao != null) {
            simulacao = null;
        }
        if (renderizador != null) {
            renderizador = null;
        }
    }

    public void executar() {
        inicializar();

        while (janela.isOpen()) {
            processarEventos();
            atualizar();
            renderizar();
        }

        finalizar();
    }
}

File: ex-14/ex-14_python/principal.py
import sys
import random

from janela_render import JanelaRender
from simulacao import Simulacao
from renderizador import Renderizador
from corpo_celeste import CorpoCeleste
from vetor2d import Vetor2D
from vetor2f import Vetor2f
from cor import Cor
from evento import Evento


def _gerador_tokens():
    for linha in sys.stdin:
        for token in linha.split():
            yield token


class Principal:
    def __init__(self):
        self.janela = JanelaRender()
        self.simulacao = None
        self.renderizador = None
        self.larguraJanela = 1200
        self.alturaJanela = 800
        self._tokens = _gerador_tokens()

    def _proximoToken(self):
        return next(self._tokens)

    def lerDouble(self):
        return float(self._proximoToken())

    def lerInt(self):
        return int(self._proximoToken())

    def inicializar(self):
        self.janela.create(self.larguraJanela, self.alturaJanela, "Simulacao Gravitacional")
        self.janela.setFramerateLimit(60)
        self.obterDadosUsuario()

    def obterDadosUsuario(self):
        print("=== SIMULACAO DE ATRACAO GRAVITACIONAL ===")
        print()

        print("Digite a largura do espaco (em metros, ex: 100e9 para 100 milhoes de km): ", end="")
        larguraEspaco = self.lerDouble()
        print("Digite a altura do espaco (em metros): ", end="")
        alturaEspaco = self.lerDouble()

        print("Digite o intervalo de tempo (em segundos, ex: 3600 para 1 hora): ", end="")
        dt = self.lerDouble()

        self.simulacao = Simulacao(larguraEspaco, alturaEspaco, dt)
        self.renderizador = Renderizador(self.janela, larguraEspaco, alturaEspaco)

        print()
        print("Deseja posicionar os corpos aleatoriamente? (1-Sim, 0-Nao): ", end="")
        aleatorio = self.lerInt()

        for i in range(2):
            print()
            print("=== CORPO " + str(i + 1) + " ===")

            if aleatorio == 1:
                posX = random.uniform(0, larguraEspaco)
                posY = random.uniform(0, alturaEspaco)
                print("Posicao aleatoria: X=" + str(posX) + " m, Y=" + str(posY) + " m")
            else:
                print("Digite a posicao X (em metros): ", end="")
                posX = self.lerDouble()
                print("Digite a posicao Y (em metros): ", end="")
                posY = self.lerDouble()

            print("Digite a velocidade X (em m/s, ex: 30000 para 30 km/s): ", end="")
            velX = self.lerDouble()
            print("Digite a velocidade Y (em m/s): ", end="")
            velY = self.lerDouble()

            print("Digite a massa (em kg, ex: 6e27): ", end="")
            massa = self.lerDouble()

            raio = 10.0 + i * 5.0
            cor = Cor.Yellow() if i == 0 else Cor.Cyan()

            posicao = Vetor2D(posX, posY)
            velocidade = Vetor2D(velX, velY)

            corpo = CorpoCeleste(posicao, velocidade, massa, raio, cor)
            self.simulacao.adicionarCorpo(corpo)

        print()
        print("Simulacao iniciada! Pressione ESC para sair.")
        print("Pressione ESPACO para pausar/continuar.")
        print("Pressione R para reiniciar.")

    def processarEventos(self):
        evento = self.janela.pollEvent()
        while evento is not None:
            if evento.isClosed():
                self.janela.close()

            keyPressed = evento.getKeyPressed()
            if keyPressed is not None:
                if keyPressed == Evento.Escape:
                    self.janela.close()
                if keyPressed == Evento.R:
                    self.finalizar()
                    self.obterDadosUsuario()

            evento = self.janela.pollEvent()

    def atualizar(self):
        if self.simulacao:
            self.simulacao.atualizar()

            for i in range(len(self.simulacao.getCorpos())):
                corpo = self.simulacao.getCorpos()[i]
                tamanhoJanela = self.janela.getSize()

                escalaX = tamanhoJanela[0] / self.simulacao.getLarguraEspaco()
                escalaY = tamanhoJanela[1] / self.simulacao.getAlturaEspaco()

                x = float(corpo.getPosicao().getX() * escalaX)
                y = float(corpo.getPosicao().getY() * escalaY)

                posicaoTela = Vetor2f(x, y)

                corpo.adicionarPontoTrajetoria(posicaoTela)

    def renderizar(self):
        if self.renderizador and self.simulacao:
            self.renderizador.renderizar(self.simulacao)

    def finalizar(self):
        if self.simulacao:
            self.simulacao = None
        if self.renderizador:
            self.renderizador = None

    def executar(self):
        self.inicializar()

        while self.janela.isOpen():
            self.processarEventos()
            self.atualizar()
            self.renderizar()

        self.finalizar()
